### PR TITLE
SI-5907, SI-5009 case-class copy defaults only for first param list

### PR DESCRIPTION
**Pull request by lrytz**
`copy` no longer returns anonymous functions if there are multiple parameter lists, reverts most of 40e7cab7a2.
Cleaned up the special type completer for copy methods.

review by @paulp

### CODE REVIEW

File: src/compiler/scala/tools/nsc/typechecker/Unapplies.scala [hunk 1]
```diff
@@ -197,42 +197,61 @@ trait Unapplies extends ast.TreeDSL
...
+   * ClassDef of the case class.
+   */
   def caseClassCopyMeth(cdef: ClassDef): Option[DefDef] = {
     def isDisallowed(vd: ValDef) = isRepeatedParamType(vd.tpt) || isByNameParamType(vd.tpt)
-    val cparamss  = constrParamss(cdef)
-    val flat      = cparamss.flatten
+    val classParamss  = constrParamss(cdef)
 
-    if (cdef.symbol.hasAbstractFlag || (flat exists isDisallowed)) None
+    if (cdef.symbol.hasAbstractFlag || classParamss.exists(_ exists isDisallowed)) None
```
Comment (paulp) [inline src/compiler/scala/tools/nsc/typechecker/Unapplies.scala]: FYI

```
mexists(classParamss)(isDisallowed)  
```

exists for this purpose - higher in both efficiency and aesthetics.  There is also mforeach, mmap, etc.

Comment (lrytz) [inline src/compiler/scala/tools/nsc/typechecker/Unapplies.scala]: right - i use mmap often, but keep forgetting about the others..

File: test/files/run/t5907.scala [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+}
+case class C8(x: Int)(y: Int, z: Int)()(implicit s: String) {
+  override def toString = s"c8: $x, $y, $z, $s"
+}
+case class C9(x: Int)(y: Int)()()(implicit s: String) {
+  override def toString = s"c9: $x, $y, $s"
+}
+case class C10[T,U <: T](x: T)(y: U)(implicit z: T) {
+  override def toString = s"c9: $x, $y, $z"
+}
```
Comment (paulp) [inline test/files/run/t5907.scala:118]: String interpolation ftw!

**Comment by paulp** [comment]
LGTpaulp.

**Pull request merged**